### PR TITLE
Added blockpool struct

### PR DESCRIPTION
**Pull request by vadlakondaswetha**
### Description
Added blockpool struct which is responsible for creating and maintaining blocks as per the user configuration.

### Link to the issue in case of a bug fix.
NA

### Testing details
1. Manual - NA
2. Unit tests - Added
3. Integration tests - NA

### CODE REVIEW

File: internal/block/block_pool_test.go [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+	require.Nil(t.T(), err)
+
+	block, err := bp.Get()
+
+	require.Nil(t.T(), err)
+	require.NotNil(t.T(), block)
+	assert.Equal(t.T(), int64(0), block.Size())
+}
+
+func (t *BlockPoolTest) TestCreateBlock() {
```
Comment (ashmeenkaur) [inline internal/block/block_pool_test.go:111]: This test is same as TestGetWhenTotalBlocksIsLessThanThanMaxBlocks
Shall we remove one?

File: internal/block/block_pool_test.go [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+
+func TestBlockPoolTestSuite(t *testing.T) {
+	suite.Run(t, new(BlockPoolTest))
+}
+
+func (t *BlockPoolTest) TestInitBlockPool() {
+	bp, err := InitBlockPool(1024, 10)
+
+	require.Nil(t.T(), err)
+	assert.NotNil(t.T(), bp)
```
Comment (ashmeenkaur) [inline internal/block/block_pool_test.go:41]: Nit:
```suggestion
	require.NotNil(t.T(), bp)
```

File: internal/block/block_pool.go [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+import (
+	"fmt"
+	"syscall"
+)
+
+// BlockPool handles the creation of blocks as per the user configuration.
+type BlockPool struct {
+	// Channel holding free blocks.
+	blocksCh chan Block
```
Comment (charith87) [inline internal/block/block_pool.go:25]: *nit = Can we rename to blocksChannel or similar like freeBlocksChannel instead to signify its usage and helps in readability.

File: internal/block/block_pool.go [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+				return b, nil
+			}
+		}
+	}
+}
+
+// createBlock creates a new block.
+func (ib *BlockPool) createBlock() (Block, error) {
+	prot, flags := syscall.PROT_READ|syscall.PROT_WRITE, syscall.MAP_ANON|syscall.MAP_PRIVATE
+	addr, err := syscall.Mmap(-1, 0, int(ib.blockSize), prot, flags)
```
[the remark this answers is not in the document]
Comment (charith87) [reply internal/block/block_pool.go:82]: +1, can we add freeBlock() function as well so its abstracted here with totalcount and other free related things.

[the remark this answers is not in the document]
Comment (vadlakondaswetha) [reply internal/block/block_pool.go:82]: Added DeAllocate method. 
syscall.Map will throw error when virtual address space is full. Avoids crashing in few scenarios.

File: internal/block/block_pool.go (hunk 3 above)
Comment (vadlakondaswetha) [reply internal/block/block_pool.go]: Done.

File: internal/block/block_pool_test.go (hunk 2 above)
Comment (vadlakondaswetha) [reply internal/block/block_pool_test.go]: Done.

File: internal/block/block_pool_test.go (hunk 1 above)
Comment (vadlakondaswetha) [reply internal/block/block_pool_test.go]: yep. Removed.

File: internal/block/block.go [hunk 5]
```diff
@@ -76,3 +79,18 @@ func (m *memoryBlock) Write(bytes []byte) error {
 func (m *memoryBlock) Reader() io.Reader {
 	return bytes.NewReader(m.buffer[0:m.offset.end])
 }
+
+func (m *memoryBlock) Deallocate() error {
```
[the remark this answers is not in the document]
Comment (vadlakondaswetha) [reply internal/block/block.go:83]: createBlock is actually creating block and returning so other methods are called using that instance. So we can't put createBlock as a method on Block interface. 

Other option is to just put the function in the block.go file and call it from blockpool.go.

[the remark this answers is not in the document]
Comment (ashmeenkaur) [reply internal/block/block.go:83]: > put the function in the block.go file and call it from blockpool.go.

I think this will be better. WDYT?

**Review by ashmeenkaur — APPROVED** [review]

**Pull request merged**